doc/design/ib-vf-configuration.md

```diff
@@ -0,0 +1,123 @@
+
+## Motivation
+We have customers using the SR-IOV operator to create IB VFs, and they need a way to automate GUID assignment,
+so that IB VFs are automatically bound to the required PKeys and no additional manual configuration is needed.
+In this use case, the GUID configuration is static and known in advance.
+Now the GUIDs are assigned by the sriov-network-config-daemon randomly.
+
+### Use Cases
+
+### Goals
+
+* Per node IB GUID configuration can be read from a static json file on the host
+* Per node IB GUID configuration is static and created in advance
+* Static config file is available at the same path on every host
+
+### Non-Goals
+
+* Dynamic GUID allocation is out of scope of this proposal
+
+## Proposal
+
+Mount a configuration file as a part of `/host` hostPath in the config daemon and read it to retrieve the GUID configuration.
+
+### Workflow Description
+
+In the [`pkg/host/internal/sriov/sriov.go:configSriovVFDevices`](https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/82a6d6fdce71bd88a0d9368fb1750488e9a8e4e2/pkg/host/internal/sriov/sriov.go#L458) read from the static IB config file
+and assign the GUIDs according to the configuration. Each PF is described by either PCI address
+or the PF GUID and has either a list or a range of VF GUIDs.
+
+1. A script creating the config file is deployed to the host and creates a static GUID config file
+2. SR-IOV network operator reads the file and assign GUIDs when IB VFs are created
+3. Users employ [PKey selector](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pull/517) in sriov-network-device-plugin to create PKey-specific resource pools
+   3.1. Note: When a VF gets assigned with GUID, it communicates with subnet manager and receives the PKey associated with that GUID. This PKey is stored in the sysfs and device-plugin reads it to populate pools.
+4. Users create an IB network CR
+5. Users create pods and request resources from the specific pools
+
+### Implementation Details/Notes/Constraints
+
+There can be fewer VFs created than GUIDs. To persist the dynamic nature of the SR-IOV Network operator,
+it’s proposed not to return an error in this case but assign as many GUIDs as possible.
+To ensure that nothing breaks when users add/remove VFs, the GUID distribution order should always be the same for each individual host.
+
+Parallel NIC configuration should be supported by making the GUID immutable. VF id should be used to determine the correct GUID for each VF.
+
+If there are fewer GUIDs than VFs, then all the GUIDs should be assigned. For the rest of VFs, apply random GUIDs as it's been done before.
+
+If config file doesn't exist in the predefined location, proceed as usual and assign random GUIDs to VFs.
+
+### Config file
+
+Config file can be stored at `/etc/sriov-operator/infiniband/guids`. This location is writable across most cloud platforms.
+
+Example of the config file:
+
+```json
+[
+  {
```

Comment:
can you add here the go type used for an entry ?

Reply:
Added a go struct declaration